core/src/com/unciv/logic/battle/Battle.kt

```diff
@@ -482,19 +483,33 @@ object Battle {
         promotions.XP += xpGained
 
         if (!otherIsBarbarian && civ.isMajorCiv()) { // Can't get great generals from Barbarians
-            val greatGeneralPointsBonus = thisCombatant
-                .getMatchingUniques(UniqueType.GreatPersonEarnedFaster, stateForConditionals, true)
-                .filter { unique ->
-                    val unitName = unique.params[0]
-                    // From the unique we know this unit exists
-                    val unit = civ.gameInfo.ruleset.units[unitName]!!
-                    unit.uniques.contains("Great Person - [War]")
-                }
-                .sumOf { it.params[1].toDouble() }
-            val greatGeneralPointsModifier = 1.0 + greatGeneralPointsBonus / 100
-
-            val greatGeneralPointsGained = (xpGained * greatGeneralPointsModifier).toInt()
-            civ.greatPeople.greatGeneralPoints += greatGeneralPointsGained
+            val greatGeneralUnits = civ.gameInfo.ruleset.units.values
+                    .filter { it.hasUnique(UniqueType.GreatPersonFromCombat, stateForConditionals) }
+                    // Check if the unit is allowed for the Civ, ignoring build constrants
+                    .filter { !it.getRejectionReasons(civ).any { reason ->
+                        reason.type != RejectionReasonType.Unbuildable &&
+                        reason.type != RejectionReasonType.CannotBeBuilt &&
+                        reason.type != RejectionReasonType.CannotBeBuiltUnhappiness &&
+                        // Allow Generals even if not allowed via tech
+                        reason.type != RejectionReasonType.RequiresTech &&
+                        reason.type != RejectionReasonType.Obsoleted }
```

Comment:
Should be something like `val ignorableRejectionReasons = setOf(...)`
And then this should be `reason,type !in ignorableRejectionReasons`